### PR TITLE
feat: add IA3 prompt tuning

### PR DESCRIPTION
**Pull request by maw501**
This PR adds $(IA)^3$ tuning per [Few-Shot Parameter-Efficient Fine-Tuning is Better and Cheaper than In-Context Learning](https://arxiv.org/pdf/2205.05638.pdf).

The main change is the addition of two new classes, `ParallelMLPIA3` and `ParallelSelfAttentionIA3`. Both of these simply create a new scaling vector then modify the forward pass to perform the (re)scaling. Note that the forward pass is otherwise unchanged except for where we apply $(IA)^3$.

This has been tested using model parallelism (i.e. tensor parallelism in DeepSpeed terminology) on the Stability cluster.

**Notes**

* There are no hyperparameters for the method.
* We rescale each attention head (the paper is unclear here) independently.

**TODO**

Pipeline parallelism is currently failing with $(IA)^3$. To replicate the issue:
* `pipe-parallel-size : 2 , model-parallel-size : 1, ia3_tuning: True`. Note: `num_gpus` not set. 👎 
* `pipe-parallel-size : 2 , model-parallel-size : 1, ia3_tuning: False`. Note: `num_gpus` not set. 👍 
image

### CODE REVIEW

File: configs/local_setup.yml [hunk 1]
```diff
@@ -26,5 +26,7 @@
   "log-dir": "logs",
   "use_wandb": True,
   "wandb_host": "https://api.wandb.ai",
-  "wandb_project": "neox"
+  "wandb_project": "neox",
+  "num_gpus": 1,
+  "ia3_prompt_tuning": True
```
Comment (jackapbutler) [inline configs/local_setup.yml:31]: I've been thinking a nice pattern for the GPT-NeoX configurations is to make them composable so you can setup different combinations using the script syntax in this [experiment script](https://github.com/OpenBioML/chemnlp/tree/main/experiments#working-with-the-stability-cluster) without having the touch the core configs of cluster / model;
* cluster config
* model config
* prompt tuning config (additional option)
* (IA)3 config (additional option)

The way soft prompt tuning is handled in GPT-NeoX has a nested dictionary of further PEFT-related options. Are there any equivalent options for around initialisation, etc (IA)3?
```
{
  # prompt tuning settings
   "soft_prompt_tuning": {
      "enabled": True, # also freezes all other parameters
      "n_tokens": 10,
      "init_string": "",
      "init_range": 0.5,
   }
}
```

Also, in theory I guess there's nothing stopping us from using both (IA)3 + prompt tuning at the same time.

Comment (MicPie) [reply configs/local_setup.yml:31]: In the paper the just initialize the IA3 vectors with 1s (https://arxiv.org/pdf/2205.05638.pdf#page=6 passage below the formula in the middle):
> lk, lv, and lff areallinitializedwithonesso that the overall function computed by the model does not change when they are added.

I'm not sure if adding some noise to the vectors could make sense? But likely not.

Comment (MicPie) [reply configs/local_setup.yml:31]: Maybe rename `ia3_prompt_tuning` to `ia3_tuning` because the call the vectors adapters and promp tuning is related but different?
But yeah, this is just a minor detail. 😃

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -590,6 +593,166 @@ def forward(self, hidden_states, attention_mask, layer_past=None):
...
+            param = torch.nn.Parameter(
+                torch.empty(
+                        self.hidden_size_per_partition,
+                        device=torch.cuda.current_device(),
+                        dtype=neox_args.params_dtype,
+                    )
+                )
+            param.model_parallel = True
+            param.partition_dim = 0
+            #param.stride = stride
```
Comment (MicPie) [inline megatron/model/transformer.py:640]: I'm just curious, did you need to set the stride?

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -590,6 +593,166 @@ def forward(self, hidden_states, attention_mask, layer_past=None):
...
+        mixed_x_layer = mixed_x_layer.view(*new_tensor_shape)
+
+        # [sq, b, np, 3 * hn] --> 3 [sq, b, np, hn]
+        (query_layer, key_layer, value_layer) = mpu.split_tensor_along_last_dim(
+            mixed_x_layer, 3
+        )
+        # Apply IA3 rescaling to keys & values:
+        def _apply_ia3_rescaling(layer, scale_vector):
+            layer_size  = layer.shape
+            layer = layer.reshape(layer_size[0], layer_size[1], -1)
```
Comment (MicPie) [inline megatron/model/transformer.py:671]: Just to think it through easier:
`[sq, b, np, hn] --> [sq, b, -1]`

From the top of the file:
```
     s: sequence length
     b: batch size
     h: hidden size
     n: number of attention heads
     p: number of model parallel partitions
     np: n/p
     hn: h/n
```

Comment (MicPie) [reply megatron/model/transformer.py:671]: With this setup we have separate vectors for each attention head and partition), right?

Comment (MicPie) [reply megatron/model/transformer.py:671]: The paper is here unclear, at least I don't find more infos when I search for "head" or "shared". But I'll have a look at the paper.

File: megatron/model/transformer.py (hunk 2 above)
Comment (MicPie) [reply megatron/model/transformer.py:640]: I'm not sure if this is used (all the time), but for the FFN the similar line is uncommented.

File: megatron/mpu/layers.py [hunk 4]
```diff
@@ -564,6 +564,71 @@ def forward(self, input_):
...
+    def forward(self, input_):
+        if self.use_mup and self.mup_rescale_parameters:
+            input_ /= self.width_mult()
+        # Set up backprop all-reduce.
+        input_parallel = copy_to_model_parallel_region(input_)
+        # Matrix multiply.
+
+        bias = self.bias if not self.skip_bias_add else None
+        output_parallel = F.linear(input_parallel, self.weight, bias)
+        output_parallel *= self.l_ff  # apply IA3 rescaling
```
Comment (MicPie) [inline megatron/mpu/layers.py:623]: In the paper they multiply the vector after the non-linearity, which I guess makes it more complex. Let me think about it and then we can also discuss it again, if you want. I commented also above where the non-linearity is applied in the FFN.

Comment (maw501) [reply megatron/mpu/layers.py:623]: No you are correct, I will change this - good catch!

File: configs/local_setup.yml (hunk 1 above)
Comment (maw501) [reply configs/local_setup.yml:31]: > The way soft prompt tuning is handled in GPT-NeoX has a nested dictionary of further PEFT-related options. Just wondering, are there any equivalent options for around initialisation, etc for (IA)3?
> 
> ```
> {
>   # prompt tuning settings
>    "soft_prompt_tuning": {
>       "enabled": True, # also freezes all other parameters
>       "n_tokens": 10,
>       "init_string": "",
>       "init_range": 0.5,
>    }
> }
> ```

There are no hyperparameters given we're not adding the extra loss terms.

File: megatron/model/transformer.py (hunk 3 above)
Comment (MicPie) [reply megatron/model/transformer.py:708]: Yes, agree, separate vectors per head are fine.
